### DISCUSSION TITLE
RESTful API architecture

### DISCUSSION THREAD

**Discussion by vcz-fr** [Azure/ipam#34] [Ideas]
Hello,

I am currently working with Infrastructure as Code tooling to cleanly integrate this IPAM into the provisioning of network components reserving address blocks. Unfortunately, the design of this API clashes with the way the IaC tool operates. I am currently using [Terraform](https://www.terraform.io/) with the [Mastercard/restapi](https://registry.terraform.io/providers/Mastercard/restapi/latest/docs) provider. That provider is picky but works until it needs to validate resource creation, and then deletion.

This is not a feature request as it requires an API redesign. However, code and OpenAPI 3 examples can be provided if needed.

### The issue

The API is designed in such a way that batch operations are possible with no single operation counterpart. Single operations are technically possible (a batch containing one operation is indeed a single operation) but not offered as primitives.

### Definitions

Skip this section if you are at ease with the topics of day one, day two operations and so on. We'll define the resource lifecycle as three operating phases: creation, operation and decommissioning.

#### Creation

We assume the resource we wish to create does not exist yet.

1. Create the resource using the Create endpoint `POST /parent/{parentId}/resource`, retrieve its unique identifier;
2. [If: The resource creation has succeeded] Check if the resource has been created using the Read endpoint `GET /parent/{parentId}/resource/{resourceId}`.

Save information returned during the resource creation and/or verification query.

#### Operation

We assume the resource is present. Either imported into the IaC tooling or created by it.

1. Check if the resource exists using the Read endpoint `GET /parent/{parentId}/resource/{resourceId}`, retrieve its current value;
2. [If: The resource exists AND is unchanged] Skip the resource update **[END]**;
3. [If: The resource exists AND is changed] Call the single operation Update endpoint with the unique ID of the resource `PUT/PATCH /parent/{parentId}/resource/{resourceId}`;
4. [If: The resource exists AND and its update succeeded] Check if the resource has been updated using the Read endpoint `GET /parent/{parentId}/resource/{resourceId}`. 

#### Decommissioning

We assume the resource is present. Either imported into the IaC tooling or created by it.

1. Check if the resource exists using the Read endpoint `GET /parent/{parentId}/resource/{resourceId}`, retrieve its current value;
2. [If: The resource does not exist] Skip the resource decommissioning **[END]**;
3. [If: The resource exists] Call the single operation Delete endpoint with the unique ID of the resource `DELETE /parent/{parentId}/resource/{resourceId}`;
4. [If: The resource deletion succeeded] Check if the resource has been deleted using the Read endpoint `GET /parent/{parentId}/resource/{resourceId}`. This invocation should fail.

### Discussion

When it comes to API design, it is common to start with primitives. Looping over primitives has the same effect as invoking a batch endpoint. There is of course network overhead in this approach but having primitives is always good from a software engineering standpoint. Additionally, batch operations make reporting statuses more challenging. For instance: what should happen when one in many Create fails? Should we report a fail or a success? Roll back the complete operation?

Batch operations come in handy when network overhead becomes non negligible. This is the case for data ingestion endpoints which may receive many concurrent updates from the same origin. If we follow RESTful with API path segments looking like `/resource1Type/{resource1Value}/resource2Type/{resource2Value}` then there should be no place for such an endpoint unless a resource cannot be named "batch". If this is guaranteed, then batch operation endpoints could be added to `/resource1Type/{resource1Value}/resource2Type/batch`.

**Note:** This is an ugly hack of commonly accepted API design practices to specifically allow for batch operations.

- Many Create: `POST /resource1Type/{resource1Value}/resource2Type/batch`;
- Many Update: `PUT/PATCH /resource1Type/{resource1Value}/resource2Type/batch`;
- Many Delete: `DELETE /resource1Type/{resource1Value}/resource2Type/batch`.

You will notice that there is no Read endpoint here. This because in the HTTP specification GET requests cannot contain a request body. If a "many Read" endpoint is needed, the operation could be seen as a query, at which point it could be implemented as:

- Many Read: `POST /resource1Type/{resource1Value}/resource2Type/query`.

### OneOf alternative design

It is tempting to accept `List(Resource) OR Resource` in the body of a single operation endpoint. However this gives an additional responsibility to the endpoint and may render implementation more complex on the API side as well as on the integration side! Typed languages require users to specify which data type they should expect and validate for, often automating that part (serializer-deserializer, often implemented as part of HTTP server framework implementations). Such a design breaks this behavior and requires users to re-validate data manually in their controllers.

---

That was a long post. Thank you for your attention and have a nice day!

**Comment by DCMattyG** [comment]
Hi @vcz-fr, thank you for the detailed notes above. I guess I'm trying to understand what exactly the ask is here. As outlined in other discussions, the IPAM tool doesn't actually manipulate Azure in any way. The tool provides visibility by READ ONLY for network related objects in Azure (vNETs, Subnets, Endpoints) and provides an overlay in the form of Spaces & Blocks. The tools also provides a way to reserve a CIDR range from a provided Block, then using Azure API's you can create a vNET from said CIDR range and the IPAM tool will automatically associate the new vNET to the target Block by way of a tag applied to the resource.

For these uses, the mechanism of interacting with IPAM in an IaC fashion is generally to request a CIDR reservation, grab the output, and then the next stages involve the Azure API directly.

So, all of that being said, can you help me understand more closely what you'd like to see from the IPAM API's and the business flow that would help enable? We're always interested in how we can improve this product for everyone, so we are happy to consider any & all input.

P.S. If you like to meeting directly and chat live, we're also happy to make the time for you if that is preferred.

Thank you very much!

**Comment by vcz-fr** [comment]
Indeed, I see that my actual request was not that clear. Sorry for that.
What I felt was missing is a way to control each API resource individually, that is using single-resource endpoints. This unlocks a simple integration path with Terraform, as long as you trust providers such as [Masercard/restapi](https://registry.terraform.io/providers/Mastercard/restapi/latest) of course.

Here is an example of how I would integrate the IPAM with VNET provisioning if I had access to single-resource endpoints:

```tf
#>> config.tf
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 3.23"
    }
    restapi = {
      source  = "Mastercard/restapi"
      version = "~> 1.17"
    }
  }
}

provider "azurerm" {
  features {}
  subscription_id = var.subscription_id
}

provider "restapi" {
  uri = "https://ipam.example.com/api"
  headers = {
    Content-Type  = "application/json"
    Authorization = "Bearer ${var.ipam_token}"
  }

  create_returns_object = true
}

#>> vars.tf
variable "ipam_token" {
  type        = string
  description = "IPAM authentication token"
  sensitive   = true
}

variable "cidr_mask" {
  type        = string
  description = "VNET CIDR mask, the actual CIDR is allocated by the IPAM"
}

variable "vnet_name" {
  type        = string
  description = "VNET name, does not include prefix"
}

variable "rg_name" {
  type        = string
  description = "Resource Group name to which the VNET should be attached"
}

variable "ipam_space" {
  type        = string
  description = "IPAM Space where the VNET address block should be allocated. Refer to the IPAM documentation."
}

variable "ipam_block" {
  type        = string
  description = "IPAM Block where the VNET address block should be allocated. Refer to the IPAM documentation."
}

#>> vnet.tf
data "azurerm_resource_group" "vnet_rg" {
  name = var.rg_name
}

resource "restapi_object" "reservation" {
  path = "/spaces/${var.ipam_space}/blocks/${var.ipam_block}/reservations"
  data = jsonencode({
    size = var.cidr_mask
  })
}

resource "azurerm_virtual_network" "vnet" {
  name                = "vnet-${var.vnet_name}"
  location            = data.azurerm_resource_group.vnet_rg.location
  resource_group_name = var.rg_name
  # This is where we retrieve the CIDR reserved by the IPAM and give it to the VNET
  address_space       = [restapi_object.reservation.api_data.cidr]
}

# This only declares the VNET on the IPAM, is there a path to associate a VNET to an IPAM reservation ID?
resource "restapi_object" "vnet_association" {
  path = "/spaces/${var.ipam_space}/blocks/${var.ipam_block}/networks"
  data = jsonencode({
    id     = azurerm_virtual_network.vnet.id
    active = true
  })
}
```

This provisions today but will cause reprovisioning every time this code is executed as we are missing a single-resource Read endpoint for Reservations and VNET associations. This is what the provisioning plan looks like:
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the
following symbols:
  + create

Terraform will perform the following actions:

  # azurerm_virtual_network.vnet will be created
  + resource "azurerm_virtual_network" "vnet" {
      + address_space       = (known after apply)
      + dns_servers         = (known after apply)
      + guid                = (known after apply)
      + id                  = (known after apply)
      + location            = "westeurope"
      + name                = "vnet-baz"
      + resource_group_name = "ipam-service-rg"
      + subnet              = (known after apply)
    }

  # restapi_object.reservation will be created
  + resource "restapi_object" "reservation" {
      + api_data        = (known after apply)
      + api_response    = (known after apply)
      + create_response = (known after apply)
      + data            = jsonencode(
            {
              + size = "24"
            }
        )
      + id              = (known after apply)
      + path            = "/spaces/FOO/blocks/bar/reservations"
    }

  # restapi_object.vnet_association will be created
  + resource "restapi_object" "vnet_association" {
      + api_data        = (known after apply)
      + api_response    = (known after apply)
      + create_response = (known after apply)
      + data            = (known after apply)
      + id              = (known after apply)
      + path            = "/spaces/FOO/blocks/bar/networks"
    }

Plan: 3 to add, 0 to change, 0 to destroy.
```

Even though Terraform has no native way to keep state on REST resources (as of 1.3), using a basic REST provider could be an elegant way to manage IPAM Reservations and IPAM VNET associations. This approach also works with the other resources exposed by the API.

**Reply by DCMattyG** [reply]
Hi @vcz-fr, one thing to note is this section is not required:

```
# This only declares the VNET on the IPAM, is there a path to associate a VNET to an IPAM reservation ID?
resource "restapi_object" "vnet_association" {
  path = "/spaces/${var.ipam_space}/blocks/${var.ipam_block}/networks"
  data = jsonencode({
    id     = azurerm_virtual_network.vnet.id
    active = true
  })
}
```

When you get the IPAM token, it also provides a TAG that you can add to the vNET when you create it. Adding this tag will automatically associate the new VNET with the target Block in IPAM. IPAM scans the vNET's once every 60 seconds to find these tags and create the new associations.

You can see what is returned at the end of this section in our documentation: https://azure.github.io/ipam/#/api/README?id=sample-api-calls

Does that help to clarify at all?

**Reply by vcz-fr** [reply]
Oh, so that was what the tag in the response was for! I mistook it for some sort of correlation ID similar to what APMs produce.

If this can help someone in the future, the Terraform code above only changes slightly:
```tf
#>> vnet.tf
data "azurerm_resource_group" "vnet_rg" {
  name = var.rg_name
}

resource "restapi_object" "reservation" {
  path = "/spaces/${var.ipam_space}/blocks/${var.ipam_block}/reservations"
  data = jsonencode({
    size = var.cidr_mask
  })
}

resource "azurerm_virtual_network" "vnet" {
  name                = "vnet-${var.vnet_name}"
  location            = data.azurerm_resource_group.vnet_rg.location
  resource_group_name = var.rg_name
  address_space       = [restapi_object.reservation.api_data.cidr]
  tags = {
    X-IPAM-RES-ID = restapi_object.reservation.id
  }
}
```

To the same readers: even though this code is correct today, an API change could cause it not to work in a future release of this project.

**Reply by DCMattyG** [reply]
You are amazing @vcz-fr, thank you SO MUCH for the above Terraform snippets. I know these will be incredibly useful to the rest of the community!

That said, does this address some of your concerns for the time being?

**Reply by vcz-fr** [reply]
Thank you very much! I too hope these can help engineers in their research.

Unfortunately, there is a caveat with the current snippet: Terraform cannot determine whether the Reservation still exists as the restapi provider invokes the `GET /api/spaces/${space}/blocks/${block}/reservations/${reservation_id}` endpoint which will always return HTTP 404 - Not Found as the path is not exposed by the API. Because of that, Terraform will always create a new IPAM Reservation on apply 😕

Semantically, that path should return information about the IPAM Reservation `${reservation_id}` belonging to the IPAM Block `${space}/${block}`. The same information that is returned when the IPAM Reservation is created.

Now, this is the complex part: are single-resource endpoints the only thing that is missing? That's why I initially started this discussion as a general API design improvement in place of a use case based change.

**Reply by DCMattyG** [reply]
> Unfortunately, there is a caveat with the current snippet: Terraform cannot determine whether the Reservation still exists as the restapi provider invokes the `GET /api/spaces/${space}/blocks/${block}/reservations/${reservation_id}` endpoint which will always return HTTP 404 - Not Found as the path is not exposed by the API. Because of that, Terraform will always create a new IPAM Reservation on apply 😕
> 
> Semantically, that path should return information about the IPAM Reservation `${reservation_id}` belonging to the IPAM Block `${space}/${block}`. The same information that is returned when the IPAM Reservation is created.

I think this is definitely something we can add in, it does indeed make sense to me.

> Now, this is the complex part: are single-resource endpoints the only thing that is missing? That's why I initially started this discussion as a general API design improvement in place of a use case based change.

We have single endpoints for most Azure IPAM constructs such as Spaces and Blocks. As you pointed out, we should indeed add one for Reservations for the reasons you stated above. Now, for the items we're actually pulling from Azure for visibility (vNETs, Subnets, Endpoints), I'm not sure that it would make sense as these are really intended for "Read Only" functions to bubble up their visibility with the tool (primarily the UI).

Would you agree with the above?

**Reply by vcz-fr** [reply]
This is a sensible approach. Resources that are proxied by the IPAM would not benefit from being fully managed by the IPAM today. Whenever this becomes the case, new API resources would absorb the responsibility.

In that sense, I fully agree with your statements regarding IPAM-managed and Azure-managed resources 👍

How do we proceed from there? Should we identify and document missing single-resources endpoints then create a GitHub issue for each of them?

**Reply by DCMattyG** [reply]
@vcz-fr, I think that's a great start! Outside of the reservations, let me know which other ones are missing, and I'll add them to our project plan ASAP for you!

**Reply by vcz-fr** [reply]
Here is what I gathered from [https://ipamdev.azurewebsites.net/api/docs](https://ipamdev.azurewebsites.net/api/docs), which looks like the the most recent development version of the service. I found that URL in the documentation.

I did not include Networks in the following tables since the IPAM acts as a proxy for those resources. If I forgot to include an endpoint feel free to reuse the tables to give you a head start.

I believe the most valuable ones to simplify integrations of tooling managing each resource individually would be the missing Single-resource endpoints. Of those, I would prioritize:

1. **Reservation**: Most valuable resource of the service;
2. **Block**: Almost complete, updating would require a delete-recreate in single-resource;
3. **Admin**: Not pressing, setup and forget;
4. **Space**: Not pressing, even more setup and forget;
5. **Exclusion**: Exclusions limit the visibility of the IPAM on the Azure networking setup.

In terms of API endpoints (Create, Read, Update,  Delete), all four have the same priority for container resources and Update can be left blank for resources that cannot be updated in place, i.e. without deleting and creating them again.

# Single-resource endpoints

| ↓ Resource type / Access →    | Create (1)                                               | Read (1)                                   | Update (1)                    | Delete (1)                                        |
|------------------------------|----------------------------------------------------------|--------------------------------------------|-------------------------------|---------------------------------------------------|
| IPAM Admin                   | **POST** /api/admin/admins                               | -                                          | -                             | **DELETE** /api/admin/admins/{objectId}           |
| Azure Subscription exclusion | -                                                        | -                                          | -                             | **DELETE** /api/admin/exclusions/{subscriptionId} |
| IPAM Space                   | **POST** /api/spaces                                     | **GET** /api/spaces/{space}                | **PATCH** /api/spaces/{space} | **DELETE** /api/spaces/{space}                    |
| IPAM Block                   | **POST** /api/spaces/{space}/blocks                      | **GET** /api/spaces/{space}/blocks/{block} | -                             | **DELETE** /api/spaces/{space}/blocks/{block}     |
| IPAM Reservation             | **POST** /api/spaces/{space}/blocks/{block}/reservations | -                                          | -                             | -                                                 |

# Multiple-resource endpoints

| ↓ Resource type / Access →   | Create (N)                                                    | Read (N)                                        | Update (N)                    | Delete (N)                                         |
|------------------------------|---------------------------------------------------------------|-------------------------------------------------|-------------------------------|----------------------------------------------------|
| IPAM Admin                   | **PUT** /api/admin/admins                                     | **GET** /api/admin/admins                       | **PUT** /api/admin/admins     | **PUT** /api/admin/admins                          |
| Azure Subscription exclusion | **POST** /api/admin/exclusions  **PUT** /api/admin/exclusions | **GET** /api/admin/exclusions                   | **PUT** /api/admin/exclusions | **PUT** /api/admin/exclusions                      |
| IPAM Space                   | -                                                             | **GET** /api/spaces                             | -                             | -                                                  |
| IPAM Block                   | -                                                             | **GET** /api/spaces/{space}/blocks              | -                             | -                                                  |
| IPAM Reservation             | -                                                             | **GET** /api/spaces/{space}/blocks/reservations | -                             | **DELETE** /api/spaces/{space}/blocks/reservations |
